### PR TITLE
Prevent duplicates in column name hints. Improve formatting.

### LINKED ISSUE #44130: Hints about column names can contain duplicates.
**Describe the issue**
```
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'total_byte' while processing query: 'SELECT name, total_bytes FROM system.tables WHERE total_byte > 1000000000 ORDER BY total_bytes DESC', required columns: 'name' 'total_bytes' 'total_byte', maybe you meant: ['name','total_bytes','total_bytes']. (UNKNOWN_IDENTIFIER)
```

The name `total_bytes` is duplicated.

The message should be written in normal English, like:

`maybe you meant 'this', 'that' or 'whatever'`
instead of
`maybe you meant ['this', 'that', 'whatever']`

### PR DESCRIPTION
**Pull request by jh0x**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Prevent duplicate column names hints. Closes #44130. 

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Interpreters/TreeRewriter.cpp [hunk 1]
```diff
@@ -1243,16 +1244,22 @@ void TreeRewriterResult::collectUsedColumns(const ASTPtr & query, bool is_select
         if (storage)
         {
             std::vector<String> hint_name{};
+            std::set<String> helper_hint_name{};
```
Comment (kssenii) [inline src/Interpreters/TreeRewriter.cpp:1247]: Let's just use only one container, there is no point in having two of them
```suggestion
            std::set<String> hint_name{};
```

File: src/IO/WriteHelpers.h [hunk 2]
```diff
@@ -1098,6 +1098,28 @@ inline String toString(const T & x)
...
+        writeChar(',', buf);
+        writeChar(' ', buf);
+    }
+    if (size >= 2)
+    {
+        writeQuoted(x[size - 2], buf);
+        writeText(finalSep, buf);
+    }
+    if (size >= 1)
+        writeQuoted(x[size - 1], buf);
```
Comment (kssenii) [inline src/IO/WriteHelpers.h:1118]: this code can be simplified
```suggestion
    for (auto it = x.begin(); it != x.end(); ++it)
    {
        if (it != x.begin())
        {
            if (std::next(it) == x.end())
                writeString(final_sep, buf);
            else
                writeString(", ", buf);
        }
        writeQuoted(x[i], buf);
    }
```

File: src/Interpreters/TreeRewriter.cpp (hunk 1 above)
Comment (jh0x) [reply src/Interpreters/TreeRewriter.cpp:1247]: I did an extra container just to check for duplicates and I didn't want to reorder the hints (as per https://github.com/ClickHouse/ClickHouse/issues/44130#issuecomment-1360066993).
Alternatively I could also do the following:
```cpp
auto hints = storage->getHints(name);
for (auto& hint : hints)
{
    if (std::find(std::begin(hint_name), std::end(hint_name), hint) == std::end(hint_name))
        hint_name.push_back(hint);
}
```

File: src/IO/WriteHelpers.h (hunk 2 above)
Comment (jh0x) [reply src/IO/WriteHelpers.h]: Thanks! Pushed the simplification.

File: src/Interpreters/TreeRewriter.cpp (hunk 1 above)
Comment (kssenii) [reply src/Interpreters/TreeRewriter.cpp:1247]: Ok, but let's add a comment that we want to preserve the order

File: src/Interpreters/TreeRewriter.cpp [hunk 3]
```diff
@@ -1243,16 +1244,22 @@ void TreeRewriterResult::collectUsedColumns(const ASTPtr & query, bool is_select
...
             for (const auto & name : columns_context.requiredColumns())
             {
                 auto hints = storage->getHints(name);
-                hint_name.insert(hint_name.end(), hints.begin(), hints.end());
+                for (auto& hint : hints)
+                {
+                    auto res = helper_hint_name.insert(hint);
+                    if (res.second)
```
Comment (kssenii) [inline src/Interpreters/TreeRewriter.cpp:1254]:
```suggestion
                    auto [_, inserted] = helper_hint_name.insert(hint);
                    if (inserted)
```

File: src/Interpreters/TreeRewriter.cpp [hunk 4]
```diff
@@ -1243,16 +1244,22 @@ void TreeRewriterResult::collectUsedColumns(const ASTPtr & query, bool is_select
         if (storage)
         {
             std::vector<String> hint_name{};
+            std::set<String> helper_hint_name{};
             for (const auto & name : columns_context.requiredColumns())
             {
                 auto hints = storage->getHints(name);
-                hint_name.insert(hint_name.end(), hints.begin(), hints.end());
+                for (auto& hint : hints)
```
Comment (kssenii) [inline src/Interpreters/TreeRewriter.cpp:1251]:
```suggestion
                for (const auto & hint : hints)
```

**Review by kssenii — APPROVED** [review]

**Pull request merged**